File: crates/agent-loop/src/loop.rs
//! The default agent-loop driver: [DefaultAgentLoop], its [AgentLoop] impl,
//! and the cooperative-cancellation helpers (cancel_set / wait_cancel).
//!
//! It drives one user turn by appending to the session log, looping over model
//! steps, dispatching tool calls through the registry, and writing the final
//! assistant message. All dependencies (llm, session, tools) are resolved
//! from the shared celestea_core::Context at turn start.

use std::io::{self, Write};
use std::sync::{Arc, Mutex};

use async_trait::async_trait;
use tokio::sync::watch;
use celestea_core::{
    AgentConfig, AgentError, AgentLoop, Content, Context, LlmService, Message, ModelRequest,
    SessionEvent, SessionService, StreamEvent, ToolCall, ToolInput, TurnOutcome, Usage,
    ToolRegistryService,
};
use futures_util::StreamExt;

use crate::context::{estimate_tokens, trim_context};
use crate::events::{EventSink, LoopEvent};

/// The default agent loop: a stateless driver (the session log owns the
/// monotonic turn id counter — P0-A), with an optional cooperative
/// cancellation signal (W191) and an optional event sink (P1, output
/// decoupling).
///
/// Cancellation is injected via DefaultAgentLoop::with_cancel as a
/// tokio::sync::watch::Receiver<bool>; when the watch value becomes true the
/// loop stops at the next await checkpoint (tokio::select!) and returns
/// gracefully. DefaultAgentLoop::new (no signal) keeps the pre-cancel behavior
/// for back-compat, so the AgentLoop trait and run_turn signature are unchanged.
///
/// The event sink is injected via DefaultAgentLoop::with_sink (or
/// with_cancel_sink for both); every LoopEvent a turn produces is delivered
/// to it instead of being printed. A None sink keeps the legacy stdout print
/// behavior, so existing callers (and the CLI --json path) are unchanged.
pub struct DefaultAgentLoop {
    config: AgentConfig,
    /// Cooperative cancel signal. None = never cancelled (back-compat).
    cancel: Option<watch::Receiver<bool>>,
    /// Event sink. None = legacy stdout printing (back-compat).
    sink: Option<EventSink>,
    /// Shared usage tracker (W220). None = usage not recorded (back-compat).
    usage: Option<Arc<UsageTracker>>,
}

/// Thread-safe cumulative usage accounting (W220).
///
/// The agent loop records every LLM stream's StreamEvent::Usage into an
/// optional tracker; the runtime reads latest()/total() to expose usage via
/// /api/status and to drive future trimming decisions. Cheap interior
/// mutability (std::sync::Mutex) keeps record() callable from the &self loop.
#[derive(Debug, Default)]
pub struct UsageTracker {
    inner: Mutex<UsageState>,
}

#[derive(Debug, Default)]
struct UsageState {
    total: Usage,
    latest: Usage,
}

impl UsageTracker {
    pub fn new() -> Self {
        Self::default()
    }

    /// Record one LLM stream's usage: adds to the cumulative total and
    /// makes it the latest observed stream.
    pub fn record(&self, u: Usage) {
        if let Ok(mut st) = self.inner.lock() {
            st.total += u;
            st.latest = u;
        }
    }

    /// The usage of the most recent LLM stream (zeroed when none yet).
    pub fn latest(&self) -> Usage {
        self.inner.lock().map(|st| st.latest).unwrap_or_default()
    }

    /// Cumulative usage across all recorded streams.
    pub fn total(&self) -> Usage {
        self.inner.lock().map(|st| st.total).unwrap_or_default()
    }
}


impl DefaultAgentLoop {
    /// Build a loop from an AgentConfig (no cancellation signal, no sink).
    pub fn new(config: AgentConfig) -> Self {
        Self::with_bindings(config, None, None, None)
    }

    /// Build a loop that stops cooperatively once the watch value reads true.
    /// The sender half is owned by the caller (e.g. hooked to ctrl_c at the
    /// composition / CLI layer).
    pub fn with_cancel(config: AgentConfig, cancel: watch::Receiver<bool>) -> Self {
        Self::with_bindings(config, Some(cancel), None, None)
    }

    /// Build a loop that routes every LoopEvent to the given sink (stream
    /// deltas and tool events are no longer printed).
    pub fn with_sink(config: AgentConfig, sink: EventSink) -> Self {
        Self::with_bindings(config, None, Some(sink), None)
    }

    /// Build a loop with both a cooperative cancel signal and an event sink.
    pub fn with_cancel_sink(
        config: AgentConfig,
        cancel: watch::Receiver<bool>,
        sink: EventSink,
    ) -> Self {
        Self::with_bindings(config, Some(cancel), Some(sink), None)
    }

    /// Build a loop with every optional binding (W220): cooperative
    /// cancellation, an event sink, and a shared usage tracker. The
    /// convenience constructors above pass None for the unbound slots, so all
    /// existing call sites keep working unchanged.
    pub fn with_bindings(
        config: AgentConfig,
        cancel: Option<watch::Receiver<bool>>,
        sink: Option<EventSink>,
        usage: Option<Arc<UsageTracker>>,
    ) -> Self {
        Self { config, cancel, sink, usage }
    }

    /// Route one LoopEvent to the injected sink, or fall back to the legacy
    /// stdout printing when no sink is installed (back-compat).
    fn emit(&self, event: LoopEvent) {
        if let Some(sink) = &self.sink {
            sink(event);
        } else {
            let _ = Self::print_legacy(&mut io::stdout(), &event);
        }
    }

    /// The legacy default printer (used when no sink is installed): Text and
    /// Thinking deltas are written straight to the writer and flushed, exactly
    /// as the loop printed before sinks existed. Written over a generic writer
    /// so tests can assert the formatting without touching real stdout.
    pub(crate) fn print_legacy(w: &mut dyn Write, event: &LoopEvent) -> io::Result<()> {
        match event {
            LoopEvent::Text(text) => write!(w, "{}", text)?,
            LoopEvent::Thinking(text) => write!(w, "[thinking] {}", text)?,
            _ => {}
        }
        w.flush()
    }
}

/// W252: persist one accumulated "continuous thinking segment" as a single
/// [SessionEvent::ThinkingDelta] (no-op when the buffer is empty).
///
/// The session log is the source of truth for replay, but thinking deltas
/// stream token-by-token: writing one jsonl row per delta would explode the
/// log. Instead the loop concatenates consecutive deltas into a buffer and
/// flushes it only at visible boundaries (Text / Done / stream end), so one
/// contiguous reasoning burst becomes exactly ONE persisted row.
fn flush_thinking(session: &SessionService, buf: &mut String) {
    if buf.is_empty() {
        return;
    }
    let text = std::mem::take(buf);
    session.append(SessionEvent::ThinkingDelta { text });
}

/// Whether the cancellation watch is set (value true). Synchronous, safe to
/// call on a shared receiver (watch::Receiver::borrow is &self).
pub(crate) fn cancel_set(rx: &watch::Receiver<bool>) -> bool {
    *rx.borrow()
}

/// Resolve as soon as cancellation is signalled (or the sender is dropped).
/// Re-checks the current value after each notification, so it can be awaited at
/// multiple checkpoints within one turn.
pub(crate) async fn wait_cancel(rx: &mut watch::Receiver<bool>) {
    if *rx.borrow() {
        return;
    }
    loop {
        if rx.changed().await.is_err() {
            return; // sender dropped: treat as no further cancellation
        }
        if *rx.borrow() {
            return;
        }
    }
}
#[async_trait]
impl AgentLoop for DefaultAgentLoop {
    async fn run_turn(&self, ctx: &Context, user_input: &str) -> Result<(), AgentError> {
        // Resolve required services from the context.
        let llm = ctx
            .get::<LlmService>()
            .ok_or_else(|| AgentError("missing LlmService in context".into()))?;
        let session = ctx
            .get::<SessionService>()
            .ok_or_else(|| AgentError("missing SessionService in context".into()))?;
        let registry = ctx
            .get::<ToolRegistryService>()
            .ok_or_else(|| AgentError("missing ToolRegistryService in context".into()))?;

        // Turn bookkeeping: the session log is the single source of truth, and
        // it owns the monotonic turn id counter, so ids stay unique across
        // loop instances (the runtime rebuilds a loop per turn) and restarts.
        let turn_id = session.next_turn_id();
        session.append(SessionEvent::TurnStart { id: turn_id.clone() });
        session.append(SessionEvent::UserMessage { text: user_input.to_string() });

        // Cooperative cancellation (W191). Clone the receiver for this turn;
        // None (the default) keeps the pre-cancel "never cancelled" behavior.
        // At each await checkpoint we select!() the real work against wait_cancel;
        // a cancel makes the loop stop and return gracefully (TurnEnd appended).
        let mut cancel_rx = self.cancel.clone();
        let mut cancel_requested = false;

        // max_steps == 0 means unlimited steps (W220); otherwise the hard cap
        // is enforced exactly as before. The loop still ends on cancellation
        // or when the model answers without tool calls.
        let mut steps_done: usize = 0;
        // P0-A real terminal states: the outcome is decided here, written to
        // TurnEnd below, and never defaults to a fake success.
        let mut outcome = TurnOutcome::Completed;
        loop {
            if self.config.max_steps > 0 && steps_done >= self.config.max_steps {
                // Budget exhausted without a final answer: NOT completed (R1).
                outcome = TurnOutcome::StepLimit;
                break;
            }
            steps_done += 1;

            // Step-level checkpoint before issuing the next model request.
            if let Some(rx) = cancel_rx.as_ref() {
                if cancel_set(rx) {
                    // Cancelled before this step: stop (TurnEnd appended below).
                    outcome = TurnOutcome::Cancelled;
                    break;
                }
            }

            // History is derived from the log, never stored separately.
            let messages = session.derive_messages();
            // Context-budget trimming (W220): when the estimated history nears
            // the configured window threshold, drop old messages and mark the
            // removal with a short system message. No-op when
            // context_window_tokens == 0 (back-compat).
            let system_tokens = estimate_tokens(&self.config.system_prompt);
            let (messages, _trim) = trim_context(
                messages,
                system_tokens,
                self.config.context_window_tokens,
                self.config.context_trim_threshold,
                self.config.context_keep_recent,
            );

            let request = ModelRequest {
                model: self.config.model.clone(),
                system: Some(self.config.system_prompt.clone()),
                messages,
                tools: registry.schemas(),
                max_tokens: None,
                temperature: None,
            };

            // Generate, then consume the stream: Text deltas go to stdout, the
            // final Done(message) is the authoritative assistant reply. A slow
            // generate()/stream stays interruptible through select!. A
            // generation failure is a terminal ERROR state with TurnEnd, not
            // a silent return (R1: 生成失败缺 TurnEnd).
            let mut stream = match cancel_rx.as_mut() {
                Some(rx) => {
                    tokio::select! {
                        r = llm.generate(request) => match r {
                            Ok(s) => s,
                            Err(e) => {
                                outcome = TurnOutcome::Error {
                                    kind: "generate".into(),
                                    message: e.0,
                                };
                                break;
                            }
                        },
                        _ = wait_cancel(rx) => {
                            // Cancelled before/during the model's first response:
                            // break out of the step loop (TurnEnd appended below).
                            outcome = TurnOutcome::Cancelled;
                            break;
                        }
                    }
                }
                None => match llm.generate(request).await {
                    Ok(s) => s,
                    Err(e) => {
                        outcome = TurnOutcome::Error {
                            kind: "generate".into(),
                            message: e.0,
                        };
                        break;
                    }
                },
            };

            let mut assistant_text = String::new();
            let mut tool_calls: Vec<ToolCall> = Vec::new();
            let mut saw_done = false;
            // W252: consecutive Thinking deltas aggregate here and are
            // flushed as ONE ThinkingDelta per contiguous burst (see
            // flush_thinking) so replay keeps the reasoning without one
            // persisted row per streamed delta.
            let mut thinking_buf = String::new();

            // Stream consumption loop. A cancel mid-stream drops the partial
            // turn (no incomplete AssistantMessage is flushed); a stream that
            // fails or ends without a terminal frame sets the matching
            // terminal state instead of pretending success.
            // Providers may stream trailing reasoning deltas AFTER the
            // finish_reason frame. Emitting Done the moment it arrives would
            // put such Thinking blocks after the final text/…/Done sequence
            // on the wire — the UI then renders "thinking below the reply".
            // Defer the Done emission until the stream truly ends, so any
            // late Thinking/Text still lands BEFORE Done.
            let mut pending_done: Option<Message> = None;
            loop {
                let next = match cancel_rx.as_mut() {
                    Some(rx) => {
                        tokio::select! {
                            ev = stream.next() => ev,
                            _ = wait_cancel(rx) => {
                                cancel_requested = true;
                                None
                            }
                        }
                    }
                    None => stream.next().await,
                };
                let Some(event) = next else { break };
                match event {
                    // Stream deltas and the final message are routed through
                    // the sink when one is installed (no direct print).
                    // A Text delta ends the current thinking segment: flush the
                    // aggregated burst BEFORE anything that follows it.
                    StreamEvent::Text(delta) => {
                        flush_thinking(&session, &mut thinking_buf);
                        self.emit(LoopEvent::Text(delta));
                    }
                    // Accumulate (concatenate) consecutive thinking deltas;
                    // they persist as one row at the next boundary.
                    StreamEvent::Thinking(delta) => {
                        thinking_buf.push_str(&delta);
                        self.emit(LoopEvent::Thinking(delta));
                    }
                    StreamEvent::Usage(u) => {
                        // Record provider-reported usage into the optional
                        // shared tracker (runtime /api/status surface).
                        if let Some(tracker) = &self.usage {
                            tracker.record(u);
                        }
                    }
                    StreamEvent::Done(message) => {
                        // Boundary flush: the thinking segment precedes the
                        // assistant text/tool calls it belongs to in the log.
                        flush_thinking(&session, &mut thinking_buf);
                        saw_done = true;
                        for content in &message.content {
                            match content {
                                Content::Text(text) => assistant_text.push_str(text),
                                Content::ToolCall(call) => tool_calls.push(call.clone()),
                            }
                        }
                        pending_done = Some(message);
                    }
                    StreamEvent::Failed { kind, message } => {
                        // Stream-level failure (truncated turn): terminal
                        // error. Flush the partial segment first so the
                        // reasoning that did stream survives in the log.
                        flush_thinking(&session, &mut thinking_buf);
                        outcome = TurnOutcome::Error { kind, message };
                        break;
                    }
                    StreamEvent::Interrupted => {
                        // The stream was torn without a terminal frame.
                        flush_thinking(&session, &mut thinking_buf);
                        outcome = TurnOutcome::Interrupted;
                        break;
                    }
                }
            }

            // Stream-end flush: covers trailing reasoning streamed AFTER the
            // terminal frame (providers do that), a thinking-only stream that
            // ended without Done, and a cancel mid-stream. It runs before any
            // AssistantMessage/ToolCall append below, so every thinking
            // segment lands in the log ahead of the step it belongs to.
            flush_thinking(&session, &mut thinking_buf);

            // Flush the deferred Done (if any) before terminal handling, so
            // late Thinking/Text emitted above precede it on the wire.
            if let Some(m) = pending_done.take() {
                self.emit(LoopEvent::Done(m));
            }

            if cancel_requested {
                outcome = TurnOutcome::Cancelled;
                break;
            }

            if !saw_done {
                // The stream ended before a terminal frame: interrupted
                // upstream / EOF without Done — a real terminal state, never
                // a fake Completed (and no empty AssistantMessage is
                // flushed). A Failed / Interrupted stream event has already
                // decided the outcome.
                if outcome == TurnOutcome::Completed {
                    outcome = TurnOutcome::Interrupted;
                }
                break;
            }

            if tool_calls.is_empty() {
                // Plain assistant reply ends the turn.
                session.append(SessionEvent::AssistantMessage { text: assistant_text });
                break;
            }

            // Tool-call step. Protocol ordering matters: every tool call in
            // this assistant turn belongs to a single assistant message, so
            // append ALL ToolCall events first, then dispatch and append each
            // ToolResult in turn. Interleaving ToolCall/ToolResult per call
            // would misrepresent the history to the model.
            for call in &tool_calls {
                session.append(SessionEvent::ToolCall {
                    id: call.id.clone(),
                    name: call.name.clone(),
                    args: call.args.clone(),
                    parent_id: None,
                });
                self.emit(LoopEvent::ToolCall {
                    id: call.id.clone(),
                    name: call.name.clone(),
                    args: call.args.clone(),
                });
            }

            // Dispatch concurrently, bounded by max_parallel_tool_calls
            // (clamped to at least 1; a 0 limit means serial). join_all
            // resolves in input order, so ToolResult events are appended in
            // the model's original call order even though the calls run in
            // parallel — the log ordering stays deterministic. A cancel during
            // a pending batch stops dispatch (dropped batch, consistent log).
            let limit = self.config.max_parallel_tool_calls.max(1);
            for batch in tool_calls.chunks(limit) {
                let dispatch = futures_util::future::join_all(batch.iter().map(|call| {
                    registry.dispatch(ToolInput {
                        call_id: call.id.clone(),
                        name: call.name.clone(),
                        args: call.args.clone(),
                    })
                }));
                let outputs = match cancel_rx.as_mut() {
                    Some(rx) => {
                        tokio::select! {
                            o = dispatch => o,
                            _ = wait_cancel(rx) => {
                                cancel_requested = true;
                                break;
                            }
                        }
                    }
                    None => dispatch.await,
                };
                for output in outputs {
                    self.emit(LoopEvent::ToolResult(output.clone()));
                    session.append(SessionEvent::ToolResult {
                        id: output.call_id,
                        value: output.value,
                        error: output.error,
                        parent_id: None,
                    });
                }
            }
            if cancel_requested {
                outcome = TurnOutcome::Cancelled;
                break;
            }
        }

        // P0-A: every started turn ends with exactly one TurnEnd carrying the
        // real terminal state (completed / cancelled / error / step_limit /
        // interrupted) — and the sink receives exactly one terminal TurnEnd
        // event, which consumers map onto their "done" envelope.
        session.append(SessionEvent::TurnEnd { id: turn_id, outcome: outcome.clone() });
        self.emit(LoopEvent::TurnEnd(outcome));
        Ok(())
    }
}

File: packages/agent-loop/src/loop.ts
/**
 * DefaultAgentLoop — port of `crates/agent-loop/src/loop.rs`.
 *
 * One turn = one user message + N model steps. Per step the loop:
 *   1. derives the model-visible history from the session log (the log is the
 *      only source of truth) and trims it to the context budget;
 *   2. asks the `Llm` seam for a stream and consumes text / thinking / usage
 *      deltas, aggregating reasoning bursts into one persisted row;
 *   3. appends the authoritative assistant reply, or dispatches the step's
 *      tool calls through the `ToolRegistry` seam (all `tool_call` rows first,
 *      then one `tool_result` per call, in model order);
 *   4. appends whatever arrived while the turn was RUNNING — a user
 *      interjection or a worker receipt — to the log at the step boundary,
 *      right before the next model call (W513), so the running turn receives it
 *      without being interrupted and without a second turn being started;
 *   5. repeats until the model answers without tool calls, the step budget is
 *      exhausted, the turn is cancelled, or the stream fails.
 *
 * Every started turn ends with EXACTLY ONE `turn_end` — in the log and on the
 * event stream, written from a single exit point — carrying one of the five
 * real terminal states: completed / cancelled / error / step_limit /
 * interrupted. A torn stream, an exhausted budget or a cancellation is never
 * reported as `completed`.
 *
 * Cancellation is cooperative over an [AbortSignal] and re-checked at every
 * await checkpoint (before generating, while streaming, between tool batches).
 * When a tool batch is abandoned, every unanswered call of the step gets a
 * synthesized cancelled `tool_result`, so the log stays protocol-valid.
 */

import {
  AgentError,
  formatInjection,
  type AgentConfig,
  type AgentLoop,
  type Context,
  type InjectionSource,
  type LoopEvent,
  type LlmStream,
  type ModelRequest,
  type ImageRef,
  type ToolCall,
  type ToolOutput,
  type ToolRegistry,
  type TurnOutcome,
} from "@celestea/core";
import { CANCELLED_BEFORE_EXECUTION, closeIterator, errorMessage, isAborted, raceAbort } from "./cancel.js";
import { estimateTokens, trimContext } from "./context-trim.js";
import { doneEvent, toolCallEvent, toolResultEvent, turnEndEvent, type EventSink } from "./events.js";
import { dispatchCall, resolveSeams, toToolInput, type Seams } from "./seams.js";
import { absorbDone, emptyStreamOutcome, terminalFromStreamEvent, type GenerateResult, type StepResult, type StreamOutcome } from "./step.js";
import { ThinkingBuffer } from "./thinking.js";
import { UsageTracker } from "./usage.js";
import {
  RETENTION_SERVICE,
  faceToolOutput,
  newStepRetention,
  retainToolResult,
  type StepRetention,
  type ToolResultRetention,
} from "./retention.js";

/** Optional collaborators of one loop instance (`with_bindings`). */
export interface AgentLoopBindings {
  /** Cooperative cancellation; absent = the turn can never be cancelled. */
  signal?: AbortSignal;
  /** Turn-event sink; absent = events are dropped (the host owns rendering). */
  sink?: EventSink;
  /** Shared usage accounting; absent = provider usage is not recorded. */
  usage?: UsageTracker;
  /** Mid-turn injection source (W513); absent = the turn takes no interjections. */
  injections?: InjectionSource;
}

/** Bound of the "do not close while a steering message waits" extension. */
export const MAX_STEER_EXTENSIONS = 8;

export class DefaultAgentLoop implements AgentLoop {
  private readonly config: AgentConfig;
  private readonly signal: AbortSignal | undefined;
  private readonly sink: EventSink | undefined;
  private readonly usage: UsageTracker | undefined;
  private readonly injections: InjectionSource | undefined;
  /** W855: resolved from the Context once per turn (null = retention off). */
  private retention: ToolResultRetention | null = null;

  constructor(config: AgentConfig, bindings: AgentLoopBindings = {}) {
    this.config = config;
    this.signal = bindings.signal;
    this.sink = bindings.sink;
    this.usage = bindings.usage;
    this.injections = bindings.injections;
  }

  /** The config this loop drives turns with. */
  get agentConfig(): AgentConfig {
    return this.config;
  }

  /** The usage tracker bound at construction, when one was provided. */
  get usageTracker(): UsageTracker | undefined {
    return this.usage;
  }

  /** AgentLoop seam: drive one turn; rejects only on a broken Context wiring. */
  async runTurn(ctx: Context, userInput: string | null, attachments?: readonly ImageRef[]): Promise<void> {
    await this.runTurnOutcome(ctx, userInput, attachments);
  }

  /**
   * W725: the EXACT request the next step would build for `ctx` — the same
   * `buildRequest` the turn uses (system prompt + post-trim derived history +
   * `registry.schemas()`), so a read-only context snapshot can never drift from
   * what the model is actually sent.
   *
   * Read-only by construction: it never appends to the log, never dispatches a
   * tool and never touches the step budget or the usage tracker.
   */
  contextSnapshot(ctx: Context): ModelRequest {
    return this.buildRequest(resolveSeams(ctx));
  }

  /** Same turn, handing the terminal state back to the caller (hosts / tests). */
  async runTurnOutcome(ctx: Context, userInput: string | null, attachments?: readonly ImageRef[]): Promise<TurnOutcome> {
    const seams = resolveSeams(ctx);
    // W855: the host provides the session-scoped retention policy; absent = off.
    this.retention = ctx.get<ToolResultRetention>(RETENTION_SERVICE) ?? null;
    // The LOG owns the monotonic turn id counter, so ids stay unique across
    // loop instances and process restarts.
    const turnId = seams.session.nextTurnId();
    seams.session.append({ type: "turn_start", id: turnId });
    // W804: the turn's own input carries this turn's attachments. The condition
    // keeps the no-attachment row byte-identical to the pre-W804 shape.
    // W855 (C8): a `null` input with no attachments writes NO row — the turn's
    // user content is whatever the drain already injected (see the AgentLoop
    // seam doc). `null` + attachments still writes the `""`-text image row.
    if (attachments !== undefined && attachments.length > 0) {
      seams.session.append({ type: "user_message", text: userInput ?? "", attachments: [...attachments] });
    } else if (userInput !== null) {
      seams.session.append({ type: "user_message", text: userInput });
    }

    let outcome: TurnOutcome = "interrupted";
    let failure: unknown;
    let failed = false;
    try {
      outcome = await this.driveSteps(seams);
    } catch (error) {
      // W813 P2-runTurnOutcome: a seam throw while deriving the request / driving
      // a step used to escape BEFORE the single TurnEnd write, leaving turn_start
      // dangling (the watchdog then keeps the worker RUNNING forever). Capture it
      // as this turn's terminal state, write the pair below, then rethrow so the
      // broken seam is still visible to the caller.
      failed = true;
      failure = error;
      outcome = { error: { kind: "generate", message: errorMessage(error) } };
    }

    // P0-A: exactly one TurnEnd per turn, log and event stream written as a pair
    // from this single exit point — reached on the throw path too.
    seams.session.append({ type: "turn_end", id: turnId, outcome });
    this.emit(turnEndEvent(outcome));
    if (failed) throw failure;
    return outcome;
  }

  /**
   * The step loop: budget -> cancel checkpoint -> one model step.
   *
   * W515 §1 invariant: a turn must NOT reach its terminal state while the
   * `next-step` lane still holds something — the message is drained and answered
   * inside THIS turn (that is what makes "closing -> inject" real). The
   * extension is bounded, and a cancelled turn still stops immediately.
   */
  private async driveSteps(seams: Seams): Promise<TurnOutcome> {
    let stepsDone = 0;
    let extensions = 0;
    for (;;) {
      // max_steps === 0 means unlimited steps (W220); a nonzero cap stops the
      // loop without a final answer, which is a step_limit, never completed.
      if (this.config.max_steps > 0 && stepsDone >= this.config.max_steps) return "step_limit";
      stepsDone += 1;
      if (isAborted(this.signal)) return "cancelled";
      const step = await this.runStep(seams);
      if (step.kind === "continue") continue;
      if (step.kind === "final" && extensions < MAX_STEER_EXTENSIONS && this.pendingSteering() > 0) {
        extensions += 1;
        continue;
      }
      return step.kind === "cancelled" ? "cancelled" : step.outcome;
    }
  }

  /** Steering messages still waiting (the close guard of W515 §1). */
  private pendingSteering(): number {
    return this.injections?.pending?.() ?? 0;
  }

  /**
   * Derive the history from the log and trim it to the context budget. A
   * `context_window_tokens` of 0 disables trimming (back-compat).
   */
  private buildRequest(seams: Seams): ModelRequest {
    const trimmed = trimContext(
      seams.session.deriveMessages(),
      estimateTokens(this.config.system_prompt),
      this.config.context_window_tokens,
      this.config.context_trim_threshold,
      this.config.context_keep_recent,
    );
    return {
      model: this.config.model,
      system: this.config.system_prompt,
      messages: trimmed.messages,
      tools: seams.registry.schemas(),
      max_tokens: null,
      temperature: null,
    };
  }

  /** Start one model response; interruptible, never throws on provider failure. */
  private async generate(seams: Seams, request: ModelRequest): Promise<GenerateResult> {
    const raced = await raceAbort(this.signal, seams.llm.generate(request));
    if (raced.outcome === "aborted") return { kind: "cancelled" };
    if (raced.outcome === "failed") {
      // Generation failure is a terminal error state with a TurnEnd (R1),
      // never a silent return.
      return { kind: "failed", outcome: { error: { kind: "generate", message: errorMessage(raced.error) } } };
    }
    return { kind: "ok", stream: raced.value };
  }

  /** One step: inject what arrived mid-turn, generate, consume, decide. */
  private async runStep(seams: Seams): Promise<StepResult> {
    this.injectPending(seams);
    const started = await this.generate(seams, this.buildRequest(seams));
    if (started.kind === "cancelled") return { kind: "cancelled" };
    if (started.kind === "failed") return { kind: "final", outcome: started.outcome };

    const thinking = new ThinkingBuffer(seams.session);
    const stream = await this.consumeStream(started.stream, thinking);
    // Stream-end flush: trailing reasoning (providers stream it AFTER the
    // finish frame), a thinking-only stream and a mid-stream cancel all persist
    // here, ahead of the appends below.
    thinking.flush();
    // The Done event is deferred to this point, so any late thinking still
    // lands before the reply on the wire.
    if (stream.doneMessage !== null) this.emit(doneEvent(stream.doneMessage));
    return this.finishStep(seams, stream);
  }

  /**
   * Consume the response stream. A cancel drops the partial turn (no
   * incomplete AssistantMessage is flushed); a failed / torn stream records the
   * matching terminal state instead of pretending success.
   */
  private async consumeStream(stream: LlmStream, thinking: ThinkingBuffer): Promise<StreamOutcome> {
    const out = emptyStreamOutcome();
    const iter = stream[Symbol.asyncIterator]();
    for (;;) {
      const next = await raceAbort(this.signal, iter.next());
      if (next.outcome === "aborted") {
        closeIterator(iter);
        out.cancelled = true;
        break;
      }
      if (next.outcome === "failed") {
        out.terminal = { error: { kind: "stream", message: errorMessage(next.error) } };
        break;
      }
      if (next.value.done === true) break;
      const event = next.value.value;
      if (event.kind === "text") {
        thinking.flush();
        this.emit({ kind: "text", delta: event.text });
      } else if (event.kind === "thinking") {
        thinking.push(event.text);
        this.emit({ kind: "thinking", delta: event.text });
      } else if (event.kind === "usage") {
        this.usage?.record(event.usage);
      } else if (event.kind === "done") {
        thinking.flush();
        absorbDone(out, event.message);
      } else {
        thinking.flush();
        out.terminal = terminalFromStreamEvent(event);
        break;
      }
    }
    return out;
  }

  /** Turn the consumed stream into the step verdict. */
  private async finishStep(seams: Seams, stream: StreamOutcome): Promise<StepResult> {
    if (stream.cancelled) return { kind: "cancelled" };
    if (!stream.sawDone) {
      // Stream ended without a terminal frame: a real terminal state, and no
      // empty AssistantMessage is flushed.
      return { kind: "final", outcome: stream.terminal ?? "interrupted" };
    }
    if (stream.toolCalls.length === 0) {
      seams.session.append({ type: "assistant_message", text: stream.assistantText });
      return { kind: "final", outcome: stream.terminal ?? "completed" };
    }
    // Deliberate divergence from the legacy loop (README §Divergences): a torn
    // stream after a done frame ends the turn instead of dispatching tools
    // under a sticky error outcome.
    if (stream.terminal !== null) return { kind: "final", outcome: stream.terminal };
    const cancelled = await this.dispatchToolCalls(seams, stream.toolCalls);
    return cancelled ? { kind: "cancelled" } : { kind: "continue" };
  }

  /**
   * Append every `tool_call` of the step first, then dispatch in batches of
   * `max_parallel_tool_calls` (clamped to >= 1) and append one `tool_result`
   * per call in model order. Returns true when a cancellation abandoned the
   * batches.
   */
  private async dispatchToolCalls(seams: Seams, calls: readonly ToolCall[]): Promise<boolean> {
    for (const call of calls) {
      seams.session.append({ type: "tool_call", id: call.id, name: call.name, args: call.args });
      this.emit(toolCallEvent(call));
    }
    const answered = new Set<string>();
    // W855 #8b: retention is skipped for read tools, so the post-execute path
    // must know which tool produced each output (the model's call order).
    const names = new Map(calls.map((call) => [call.id, call.name] as const));
    const limit = Math.max(1, this.config.max_parallel_tool_calls);
    // W855: ONE cumulative budget per step, debited in model order.
    const step = newStepRetention();
    let cancelled = false;
    for (let start = 0; start < calls.length; start += limit) {
      const batch = calls.slice(start, start + limit);
      const raced = await raceAbort(this.signal, this.dispatchBatch(seams.registry, batch));
      // Unreachable: dispatchCall is total, so a batch never rejects.
      if (raced.outcome === "failed") throw new AgentError(`tool dispatch failed: ${errorMessage(raced.error)}`);
      if (raced.outcome === "aborted") {
        cancelled = true;
        break;
      }
      for (const output of raced.value)
        await this.recordToolResult(seams, output, answered, step, names.get(output.call_id) ?? null);
    }
    if (cancelled) this.synthesizeCancelledResults(seams, calls, answered);
    return cancelled;
  }

  /** Dispatch one batch concurrently; results keep the model's call order. */
  private dispatchBatch(registry: ToolRegistry, batch: readonly ToolCall[]): Promise<ToolOutput[]> {
    return Promise.all(batch.map((call) => dispatchCall(registry, toToolInput(call))));
  }

  private async recordToolResult(
    seams: Seams,
    output: ToolOutput,
    answered: Set<string>,
    step: StepRetention,
    toolName: string | null,
  ): Promise<void> {
    // W855 (B6): the LOG keeps the ORIGINAL value + a `surface` descriptor; the
    // MODEL/SSE face is rendered from it (retention's bounded window, or a
    // tool-authored truncation note). A failed spill keeps the original inline.
    // W855 #8b: `toolName` lets the policy exempt read tools (no read loop).
    const faces =
      this.retention === null ? faceToolOutput(output) : await retainToolResult(output, this.retention, step, toolName);
    this.emit(toolResultEvent(faces.face));
    answered.add(faces.logged.call_id);
    seams.session.append({
      type: "tool_result",
      id: faces.logged.call_id,
      value: faces.logged.value,
      error: faces.logged.error,
      ...(faces.logged.surface === undefined ? {} : { surface: faces.logged.surface }),
    });
  }

  /**
   * W267: a cancel mid-dispatch drops the in-flight batch and every later
   * batch, which would leave the assistant `tool_calls` dangling — an
   * OpenAI-compatible upstream rejects that history with 400. One synthesized
   * cancelled result per unanswered call keeps the LOG protocol-valid: in the
   * model's call order, after every real result and before TurnEnd.
   */
  private synthesizeCancelledResults(seams: Seams, calls: readonly ToolCall[], answered: ReadonlySet<string>): void {
    for (const call of calls) {
      if (answered.has(call.id)) continue;
      const error = CANCELLED_BEFORE_EXECUTION;
      seams.session.append({ type: "tool_result", id: call.id, value: null, error });
      this.emit(toolResultEvent({ call_id: call.id, value: null, render: null, error, decision: null }));
    }
  }

  /**
   * W513 step boundary: append every message that arrived while the turn was
   * running as a `user_message` row, in arrival order, before the model call
   * that follows. Returns how many rows were appended.
   *
   * The log is the only source of truth, so the injected text is part of the
   * derived history of THIS turn and of every later step of it — and it is
   * written by the same append path as the turn's own input.
   */
  private injectPending(seams: Seams): number {
    const pending = this.injections?.drain() ?? [];
    for (const injection of pending) {
      seams.session.append({ type: "user_message", text: formatInjection(injection) });
    }
    return pending.length;
  }

  /** Route one turn event to the sink; without a sink the host renders nothing. */
  private emit(event: LoopEvent): void {
    this.sink?.(event);
  }
}
